### DISCUSSION TITLE
How to enable triggers that reference other DBsets with triggers when using IDbContextFactory for context creation?

### DISCUSSION THREAD

**Discussion by gkniffin** [koenbeuk/EntityFrameworkCore.Triggered#104] [Q&A] [answered]
If using services.AddDbContextFactory<AppDbContext>(...) as in your BlazorTests project, how does one create a BeforeSave trigger which needs to create/add entities outside of the current ITriggereContext as in the SignupToMandatoryCourses trigger example?  Is there a way to reference the underlying DbContext within the trigger rather than constructor injection?

My dbcontext registration is:
`services.AddDbContextFactory<AppDbContext>(opts => ...);`

And my trigger is defined as:
```csharp
public class AddAuditTrigger : IBeforeSaveTrigger<AppUser>
    {
        private readonly IDbContextFactory<AppDbContext> _appDbContextFactory;

        public AddAuditTrigger (
            IDbContextFactory<AppDbContext> appDbContextFactory)
        {
            _appDbContextFactory = appDbContextFactory ?? throw new ArgumentNullException(nameof(appDbContextFactory));
        }

        public async Task BeforeSave(
            ITriggerContext<AppUser> context,
            CancellationToken cancellationToken)
        {

            if (context.ChangeType == ChangeType.Added)
            {
                var db = _appDbContextFactory.CreateDbContext(); //<-- This isn't the same context, so how to persist new entities on original SaveChanges()?

                db.UserAudits.Add(new EntityChange { EntityType = "SomeEntity" });

                // If I call db.SaveChanges(), it works.  But if the original SaveChanges() fails, my db is left in an inconsistent state.
                // Is this a job for transactions or is there an easier way?
            }
        }
    }
```

My triggers were working when I registered my DBContext via "services.AddDbContext<T>(...)", but when switching to the recommended "services.AddDbContextFactory<T>(...)" for Blazor Server I've hit a problem with the more complex triggers that write to entities outside of the current ITriggereContext.

**Comment by koenbeuk** [comment]
Thanks for brining this up, this is indeed a bug in the sample. What we should have used here is `services.AddTriggeredDbContextFactory<AppDbContext>` https://github.com/koenbeuk/EntityFrameworkCore.Triggered/blob/bfa1085cbf094bb7bfc9efc6195fb1080b7b65a5/src/EntityFrameworkCore.Triggered/Extensions/ServiceCollectionExtensions.cs#L61

Once using that, things should just work.

This issue is really with EF itself where AddDbContext would resolve a DbContext with the ApplicationServiceProvider scoped to the current request whereas AddPooledDbContext or AddDbContextFactory would set the ApplicationServiceProvider to the root ServiceProvider (Which is not scoped and therefore does not share your service instances). 

In short, use: `AddTriggeredDbContextFactory` and that will function as a workaround. We should fix this in the sample, feel free to create a new issue, I will update the sample at the earliest convenience.

The EF issue is tracked over here:
https://github.com/dotnet/efcore/issues/13540

**Reply by gkniffin** [reply]
Thanks for taking a look at this.  I am really loving the library so far as I am able to write cleaner code/triggers when entities are saved.

I have tried `services.AddTriggeredDbContextFactory<ApplicationDbContext>(...)` in startup.cs as suggested and I am seeing the same result.  Because the trigger calls `_contextFacotry.CreateDbContext()` to interact with other entities, the scope is different and unless I call `db.SaveChanges()` the new entities are never persisted.  

I could not find a sample where both DbContextFactory was used in service registration AND the trigger writes to other entities outside of the current context was used.  I am uncertain if this is an issue within the library or if I simply do not know how to create triggers working with other entities while using DbContextFactory registration.

**Reply by koenbeuk** [reply]
I see (have not used this myself with Blazor Server before). Indeed, since the DbContext is resolved through a Factory, this DbContext is then not registered with the DI container and hence can't be directly resolved in a Trigger (at the moment).

`services.AddTriggeredDbContextFactory` is still important here as it ensures that triggers are being resolved using your scoped IServiceProvider. Using `AddDbContextFactory` (non Triggered version) will still limit your triggers to using your root ServiceProvider.

I will create a new issue for this and work on a fix (Should be relatively trivial)

**Comment by koenbeuk** [comment, marked as answer]
This is now fixed on master through this PR: https://github.com/koenbeuk/EntityFrameworkCore.Triggered/pull/106/files

Make sure to use services.AddTriggeredDbContextFactory as well as version 2.3.1 or up (going to release that version shortly)

**Reply by gkniffin** [reply]
I am walking in circles here.  I have the AddTriggeredDbContextFactory<MyDbContext>, but I am still unable to resolve the current MyDbContext instance within the trigger to manage other entities within the same operation. In my sample, I'm trying to write to an "CountAudit" table each time a new "Count" is added.  An example of this in the "4 - BlazorTests" project while using DbContextFactory would be helpful. 

The "3 -StudentManagerAspNetCore" has an example of writing to other contexts in the SignupToMandatoryCourses trigger, but this is using AddTriggeredDbContextPool() which reuses the same context.  I need an example while using DbContextFactory instead.

**Reply by koenbeuk** [reply]
Are you sure you are on the latest 2.3.1 release? You can adjust the "4 - BlazorTests" example a bit. Simply add a constructor to the `SetCreatedOn` trigger that accepts the ApplicationDbContext and take a note of the ContextId. This should be the same as the one that was used to invoke this trigger:

![image](https://user-images.githubusercontent.com/2912652/119376842-03aff180-bcef-11eb-9fa5-6d5971dafd6b.png)

If that still does not work for you then perhaps can you share a reproducible example?

**Reply by gkniffin** [reply]
Two things:

1. The download src of Triggered was updated, but the Blazor project was still referencing 2.3.0
2. In my own project where I am using tiggers, I was trying to inject IDbContextFactory instead of the DbContext.  

Thanks for this screenshot and helping me find a solution to this.  Can I buy you a cup of coffee?
